In [38]:
import pandas as pd 
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import mean_squared_error
import numpy as np 

In [39]:
df= pd.read_csv('/datasets/users_behavior.csv')
df.info()
print(df.head(10))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB
   calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0
5   58.0   344.56      21.0  15823.37         0
6   57.0   431.64      20.0   3738.90         1
7   15.0   132.40       6.0  21911.60         0
8    7.0    43.39       3.0   2538.67         1
9   90.0   665.41      38.0  17358.61         0


In [40]:
features = df.drop(['is_ultra'], axis = 1)
target = df["is_ultra"]
features_temp, features_test, target_temp, target_test = train_test_split(features, target, test_size = 0.2, random_state = 54321)
features_train, features_valid,target_train, target_valid = train_test_split(features_temp, target_temp, test_size = 0.25, random_state =54321 )


In [46]:
for depth in [5,10, 15] :
    model_1 = DecisionTreeClassifier(random_state= 54321, max_depth = depth)
    model_1.fit(features_train, target_train)
    model_1_prediction = model_1.predict(features_valid)
    accuracy_model_1 = accuracy_score(target_valid,model_1_prediction)
    print(f"depht {depth} ,porcentaje de acierto de : {accuracy_model_1}")

depht 5 ,porcentaje de acierto de : 0.8180404354587869
depht 10 ,porcentaje de acierto de : 0.7978227060653188
depht 15 ,porcentaje de acierto de : 0.7853810264385692


de este modelo, el arbol de decision con el depth de 5 tuvo el mayor porcentaje de exactitud con un 81.8% 

In [42]:
for estimators in [10, 50, 100] :
    model_2 = RandomForestClassifier(random_state = 54321, n_estimators = estimators)
    model_2.fit(features_train, target_train)
    model_2_prediction =model_2.predict(features_valid)
    accuracy_model_2 = accuracy_score(target_valid, model_2_prediction)
    print(f"n_estimators ={estimators}, accuracy ={accuracy_model_2}")

n_estimators =10, accuracy =0.8258164852255054
n_estimators =50, accuracy =0.8320373250388803
n_estimators =100, accuracy =0.8273716951788491


In [43]:
model_3 = LogisticRegression()
model_3.fit(features_train, target_train)
model_3_prediction = model_3.predict(features_valid)
accuracy_model_3 = accuracy_score(target_valid, model_3_prediction)
print('Accuracy = ', accuracy_model_3)

Accuracy =  0.7340590979782271


Tal como se puede observar el modelo con mayor exactitud es el modelo de RandomForestClassifier cuando n estimators = 50 dando un 83% de exactitud

In [44]:
best_model = RandomForestClassifier(random_state = 54321, n_estimators = 50)
best_model.fit(features_train, target_train)
best_model_prediction = best_model.predict(features_test)
accuracy_best_model= accuracy_score (target_test , best_model_prediction)
print("la exactitud del modelo es de ",accuracy_best_model  )

la exactitud del modelo es de  0.7869362363919129


In [45]:
print('Clases unicas', target_test.unique() )
print('Cantidad de clases :', len(target_test.unique()))
random_predictions = np.random.choice(target_test.unique(), size = len(target_test))
accuracy_random = accuracy_score(target_test, random_predictions)
print("Si hicieramos aleatoriamente la prediccion, su exactitud seria del ",accuracy_random)

Clases unicas [1 0]
Cantidad de clases : 2
Si hicieramos aleatoriamente la prediccion, su exactitud seria del  0.48055987558320373


haciendo la puntuacion de la exactitud con los datos de la prueba, se confirmo que cumple con el umbral antes establecido de mas de 75% de exactitud con un total de 78.7% de exactitud, superando la prueba de cordura por 30 puntos porcentuales  , demostrando que este modelo es el de mayor exactitud 